### PR TITLE
Story 620

### PR DESCRIPTION
**Pull request by deepakr123**

### CODE REVIEW

File: module/mediaFiles/helper.js [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+     * @param {String} name - emoji name 
+     * @returns {JSON} - emoji unicode.
+     */
+
+    static getEmoji(name= "") {
+        return new Promise(async (resolve, reject) => {
+            try {
+                
+                let queryObject = {
+                    status: "active"
```
Comment (aks30) [inline module/mediaFiles/helper.js:185]: @deepakr123 - Add type filter of emoji

File: module/mediaFiles/helper.js [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+     * @param {String} name - gesture name
+     * @returns {JSON} - gesture unicode.
+     */
+
+    static getGesture(name= "") {
+        return new Promise(async (resolve, reject) => {
+            try {
+
+                let queryObject = {
+                    status: "active"
```
Comment (aks30) [inline module/mediaFiles/helper.js:243]: @deepakr123 - Add type filter of gesture

File: module/pollSubmissions/helper.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+                if (Object.keys(responseObject).length == 0) {
+                    throw new Error (messageConstants.apiResponses.RESPONSE_OBJECT_REQUIRED_CHECK)
+                }
+
+                if (userId == "") {
+                    throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK)
+                }
+
+                let pollSubmissionDocument = await this.pollSubmissionDocuments
```
Comment (aks30) [inline module/pollSubmissions/helper.js:92]: @deepakr123 - How will the userId be matched?

File: module/mediaFiles/helper.js (hunk 1 above)
Comment (deepakr123) [reply module/mediaFiles/helper.js:185]: resolved

File: module/mediaFiles/helper.js (hunk 2 above)
Comment (deepakr123) [reply module/mediaFiles/helper.js:243]: resolved

File: .env.sample [hunk 4]
```diff
@@ -130,3 +130,5 @@ ELASTICSEARCH_ENTITIES_INDEX = ""
 PUBLIC_FOLDER_PATH = "public"
 USE_USER_ORGANISATION_ID_FILTER = ""
 
+POLL_LINK_BASE_URL = ""
```
Comment (aks30) [inline .env.sample:133]: @deepakr123 - Apps portal base url

File: migrations/20200827163426-upload-app-details.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    const kendraServiceHelper = require("../generics/services/kendra");
+
+    global.migrationMsg = "Upload app details"
+    // return await db.collection('albums').updateOne({artist: 'The Beatles'}, {$set: {blacklisted: true}});
+
+    let appDetails = [
+      {
+        name: "samiksha",
+        displayName: "Samiksha",
+        appDescrption: "Get the app to discover more",
```
Comment (aks30) [inline migrations/20200827163426-upload-app-details.js:13]: @deepakr123 - Change it to description and logo only, please.

File: migrations/20200827163426-upload-app-details.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      await kendraServiceHelper.upload(
+        `public/assets/apps/${appDetails[app].appLogo}`,
+        `static/apps/${appDetails[app].appLogo}`
+      );
+
+      appDetails[app].appLogo = "static/apps/" + appDetails[app].appLogo; 
+    }
+
+    await db.collection('apps').insertMany(appDetails);
+    await db.collection('apps').createIndex({ name : 1 });
```
Comment (aks30) [inline migrations/20200827163426-upload-app-details.js:92]: @deepakr123 - This is not required.

File: migrations/20200827163426-upload-app-details.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    for( let app = 0 ; app < appDetails.length; app++ ) {
+      await kendraServiceHelper.upload(
+        `public/assets/apps/${appDetails[app].appLogo}`,
+        `static/apps/${appDetails[app].appLogo}`
+      );
+
+      appDetails[app].appLogo = "static/apps/" + appDetails[app].appLogo; 
+    }
+
+    await db.collection('apps').insertMany(appDetails);
```
Comment (aks30) [inline migrations/20200827163426-upload-app-details.js:91]: @deepakr123 - Use create API of apps in Kendra to create an app.

File: .env.sample (hunk 4 above)
Comment (deepakr123) [reply .env.sample:133]: resolved

File: module/pollSubmissions/helper.js (hunk 3 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js]: resolved

File: migrations/20200827163426-upload-app-details.js (hunk 5 above)
Comment (deepakr123) [reply migrations/20200827163426-upload-app-details.js:13]: resolved

File: migrations/20200827163426-upload-app-details.js (hunk 6 above)
Comment (deepakr123) [reply migrations/20200827163426-upload-app-details.js:92]: resolved

File: migrations/20200827163426-upload-app-details.js (hunk 7 above)
Comment (deepakr123) [reply migrations/20200827163426-upload-app-details.js]: resolved

File: module/pollSubmissions/helper.js [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+            if(userId == ""){
+                throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK);
+            }
+
+            let sortObject = {
+                createdAt: 1
+            }
+
+            let getPollList = [
+                pollsHelper.pollDocuments
```
Comment (aks30) [inline module/pollSubmissions/helper.js:200]: @deepakr123 - Use existing helper function in polls module.

File: module/pollSubmissions/helper.js [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+            if (userId == "") {
+                throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK)
+            }
+
+            let pollDocument = await pollsHelper.pollDocuments
+            (
+                { _id: pollId },
+                [
+                    "status",
+                    "endDate"
```
Comment (aks30) [inline module/pollSubmissions/helper.js:290]: @deepakr123 - We should use isDeleted filter?

Comment (deepakr123) [reply module/pollSubmissions/helper.js:290]: resolved

File: module/pollSubmissions/helper.js (hunk 8 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:200]: resolved

File: module/pollSubmissions/helper.js [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+                },
+                ["_id"]
+            )
+             
+            if (pollSubmissionDocument.length > 0) {
+                result.pollSubmissionDocument = pollSubmissionDocument
+            }
+            
+            return resolve({
+                success: true,
```
Comment (aks30) [inline module/pollSubmissions/helper.js:310]: @deepakr123 - Message is missing.

File: module/pollSubmissions/helper.js (hunk 9 above)
Comment (aks30) [reply module/pollSubmissions/helper.js:279]: @deepakr123 - Click on resolve buttons always.

File: module/pollSubmissions/helper.js [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+                throw new Error(messageConstants.apiResponses.POLL_NOT_FOUND)
+            }
+
+            let result = {
+                pollStatus: pollDocument[0].status,
+                pollSubmissionDocument: []
+            }
+
+            if (new Date() > new Date(pollDocument[0].endDate)) {
+                result.pollStatus = messageConstants.apiResponses.INACTIVE_STATUS
```
Comment (aks30) [inline module/pollSubmissions/helper.js:293]: @deepakr123 - Don't you have to update the poll document to make it inactive?

File: module/pollSubmissions/helper.js [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+                throw new Error(messageConstants.apiResponses.POLL_ID_REQUIRED_CHECK)
+            }
+
+            if (userId == "") {
+                throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK)
+            }
+
+            let pollDocument = await pollsHelper.pollDocuments
+            (
+                { _id: pollId,
```
Comment (aks30) [inline module/pollSubmissions/helper.js:275]: @deepakr123 - Should not there be a status filter here?

File: module/pollSubmissions/helper.js [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+                    throw new Error (messageConstants.apiResponses.RESPONSE_OBJECT_REQUIRED_CHECK)
+                }
+
+                if (userId == "") {
+                    throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK)
+                }
+
+                let validatePollSubmissionResult = await this.validatePollSubmission(pollId, userId);
+                
+                if (validatePollSubmissionResult.success) {
```
Comment (aks30) [inline module/pollSubmissions/helper.js:104]: @deepakr123 - Make the check if success and data false, throw an error with whatever the reason returned from validatePollSubmission function.

File: module/pollSubmissions/helper.js [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+
+                if (!result.length) {
+                    throw new Error(messageConstants.apiResponses.POLL_NOT_FOUND)
+                }
+
+                return resolve({
+                    success: true,
+                    message: messageConstants.apiResponses.POLLS_LIST_FETCHED,
+                    data: _.uniqWith(result, _.isEqual)
```
Comment (aks30) [inline module/pollSubmissions/helper.js:237]: @deepakr123 - How does this isEqual work - https://lodash.com/docs/4.17.15#isEqual as per the link it has to be passed 2 arguments?

File: module/pollSubmissions/helper.js [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+    /**
+    * Validate poll submission.
+    * @method
+    * @name validatePollSubmission
+    * @param {String} pollId - pollId
+    * @param {String} userId - userId
+    * @returns {String} - poll status and poll submission document.
+    */
+
+   static validatePollSubmission(pollId= "", userId= "") {
```
Comment (aks30) [inline module/pollSubmissions/helper.js:261]: @deepakr123 - This function should just return whether or not submission can be made or not, if yes the message, and if no then what is the reason.

File: module/polls/helper.js [hunk 16]
```diff
@@ -0,0 +1,631 @@
...
+
+                pollData.questions.forEach ( question => {
+                    
+                    let options = [];
+                    question.qid = gen.utils.generateUUId();
+                    result[question.qid] = {}
+
+                    if (question.options.length > 0) {
+                        let i = 1;
+                        question.options.forEach( option => {
```
Comment (aks30) [inline module/polls/helper.js:197]: @deepakr123 - You have to use for loop here as in some case ++i won't help as the functions are executed in parallel.

File: module/polls/helper.js [hunk 17]
```diff
@@ -0,0 +1,631 @@
+/**
+ * name : polls/helper.js
+ * author : Deepa
+ * created-date : 01-Aug-2020
+ * Description : Polls helper functionality.
+ */
+
+// Dependencies
+const formsHelper = require(MODULES_BASE_PATH + "/forms/helper");
+const appsPortalBaseUrl = process.env.APP_PORTAL_BASE_URL ? process.env.APP_PORTAL_BASE_URL : "https://apps.shikshalokam.org/";
```
Comment (aks30) [inline module/polls/helper.js:10]: @deepakr123 - Check exists and not blank condition.

File: module/polls/helper.js [hunk 18]
```diff
@@ -0,0 +1,631 @@
...
+                    throw new Error (messageConstants.apiResponses.APP_NAME_FIELD_REQUIRED);
+                }
+
+                let pollDocument = {
+                    name: pollData.name,
+                    creator: pollData.creator,
+                    createdBy: userId,
+                    startDate: new Date(),
+                    endDate: new Date(new Date().setDate(new Date().getDate() + pollData.endDate)),
+                    metaInformation: pollData.metaInformation,
```
Comment (aks30) [inline module/polls/helper.js:179]: @deepakr123 - How is it determined what goes into metaInformation ?

File: module/polls/helper.js [hunk 19]
```diff
@@ -0,0 +1,631 @@
...
+                )
+
+                if (!pollsList.length) {
+                    throw new Error(messageConstants.apiResponses.POLL_NOT_FOUND)
+                }
+
+                return resolve({
+                    success: true,
+                    message: messageConstants.apiResponses.POLLS_LIST_FETCHED,
+                    data: pollsList
```
Comment (aks30) [inline module/polls/helper.js:284]: @deepakr123 - Data should be returned as false if no data found.

File: module/polls/helper.js [hunk 20]
```diff
@@ -0,0 +1,631 @@
...
+        return new Promise(async (resolve, reject) => {
+            try {
+
+                if(pollId == ""){
+                    throw new Error(messageConstants.apiResponses.POLL_ID_REQUIRED_CHECK);
+                }
+
+                await database.models.polls.updateOne
+                (
+                    { _id: pollId },
```
Comment (aks30) [inline module/polls/helper.js:316]: @deepakr123 - Check only the creator of the poll can mark his pole as deleted or inactive etc.

File: module/polls/helper.js [hunk 21]
```diff
@@ -0,0 +1,631 @@
...
+            }
+            
+            let groupByQuestionId = _.keyBy(pollDocument[0].questions, 'qid');
+            let result = pollDocument[0].result;
+
+            Object.keys(result).forEach( singleQuestion => {
+                
+                let groupOptionsByValue = _.keyBy(groupByQuestionId[singleQuestion].options,"value");
+
+                report.title.text = groupByQuestionId[singleQuestion].question;
```
Comment (aks30) [inline module/polls/helper.js:554]: @deepakr123 - How will the report title be set if more than 1 question is present ?

File: module/pollSubmissions/helper.js (hunk 11 above)
Comment (aks30) [inline module/pollSubmissions/helper.js:293]: @deepakr123 - Date condition can be checked in query object itself.

File: module/pollSubmissions/helper.js [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+            
+            if (pollId == "") {
+                throw new Error(messageConstants.apiResponses.POLL_ID_REQUIRED_CHECK)
+            }
+
+            if (userId == "") {
+                throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK)
+            }
+
+            let pollDocument = await pollsHelper.pollDocuments
```
Comment (aks30) [inline module/pollSubmissions/helper.js:273]: @deepakr123 - Suggest creating a helper function in polls which tells given a pollId whether or not it is still open for submission or not?

File: controllers/v1/pollsController.js [hunk 23]
```diff
@@ -0,0 +1,565 @@
...
+    /**
+    * Get the poll questions
+    * @method
+    * @name getPollQuestions
+    * @param {Object} req -request Data.
+    * @param {String} req.params._id - pollId.  
+    * @returns {JSON} - poll questions and options
+    */
+
+   getPollQuestions(req) {
```
Comment (aks30) [inline controllers/v1/pollsController.js:387]: @deepakr123 - In this function, we should return a shareable URL as well. Also, create a new API which given a poll id gives a shareable link of the same.

File: module/pollSubmissions/helper.js (hunk 10 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:310]: resolved

File: module/pollSubmissions/helper.js (hunk 12 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:275]: resolved

File: module/polls/helper.js (hunk 17 above)
Comment (deepakr123) [reply module/polls/helper.js:10]: resolved

File: module/polls/helper.js (hunk 16 above)
Comment (deepakr123) [reply module/polls/helper.js:197]: resolved

File: module/polls/helper.js (hunk 20 above)
Comment (deepakr123) [reply module/polls/helper.js:316]: resolved

File: module/polls/helper.js (hunk 16 above)
Comment (deepakr123) [reply module/polls/helper.js:197]: resolved

File: module/pollSubmissions/helper.js (hunk 11 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:293]: resolved

File: module/pollSubmissions/helper.js (hunk 13 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:104]: resolved

File: module/polls/helper.js (hunk 21 above)
Comment (deepakr123) [reply module/polls/helper.js:554]: resolved

File: module/pollSubmissions/helper.js (hunk 15 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:261]: resolved

File: module/pollSubmissions/helper.js (hunk 22 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:273]: resolved

File: module/pollSubmissions/helper.js (hunk 11 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:293]: resolved

File: module/polls/helper.js (hunk 19 above)
Comment (deepakr123) [reply module/polls/helper.js:284]: resolved

File: module/polls/helper.js (hunk 18 above)
Comment (deepakr123) [reply module/polls/helper.js:179]: resolved

File: controllers/v1/pollsController.js (hunk 23 above)
Comment (deepakr123) [reply controllers/v1/pollsController.js:387]: resolved

File: module/polls/helper.js [hunk 24]
```diff
@@ -0,0 +1,745 @@
...
+
+                if(pollId == ""){
+                    throw new Error(messageConstants.apiResponses.POLL_ID_REQUIRED_CHECK)
+                }
+
+                if(userId == ""){
+                    throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK)
+                }
+
+                let pollDocument = await this.pollDocuments
```
Comment (aks30) [inline module/polls/helper.js:324]: @deepakr123 - It has to be part of updatePollDocument query itself

File: module/polls/helper.js [hunk 25]
```diff
@@ -0,0 +1,745 @@
...
+                let result = {};
+
+                for(let question=0; question< pollData.questions.length; question++) {
+   
+                    let options = [];
+                    pollData.questions[question].qid = gen.utils.generateUUId();;
+                    result[pollData.questions[question].qid] = {}
+               
+                    if (pollData.questions[question].options.length > 0) {
+                       let i = 1;
```
Comment (aks30) [inline module/polls/helper.js:205]: @deepakr123 - No need of separate variable.

File: module/polls/helper.js [hunk 26]
```diff
@@ -0,0 +1,745 @@
...
+               
+                       for(let option=0; option< pollData.questions[question].options.length; option++) {
+                           options.push(
+                                          {
+                                            value: "R" + i,
```
Comment (aks30) [inline module/polls/helper.js:210]: @deepakr123 - Instead of i, just user option+1

File: module/polls/helper.js [hunk 27]
```diff
@@ -0,0 +1,745 @@
...
+                }
+                else {
+                
+                let result = {
+                    pollId: pollQuestions[0]._id,
+                    questions: pollQuestions[0].questions,
+                    submissionId: ""
+                }
+           
+                let pollSubmissionDocument = await database.models.pollSubmissions.find
```
Comment (aks30) [inline module/polls/helper.js:470]: @deepakr123 - Use a helper function of pollSubmissions module.

File: module/pollSubmissions/helper.js [hunk 28]
```diff
@@ -0,0 +1,315 @@
...
+            
+            if (pollId == "") {
+                throw new Error(messageConstants.apiResponses.POLL_ID_REQUIRED_CHECK)
+            }
+
+            if (userId == "") {
+                throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK)
+            }
+
+            let pollDocument = await pollsHelper.validatePoll
```
Comment (aks30) [inline module/pollSubmissions/helper.js:277]: @deepakr123 - Rename validatePoll to isPollOpenForSubmission

File: module/pollSubmissions/helper.js [hunk 29]
```diff
@@ -0,0 +1,315 @@
...
+    /**
+    * Validate poll submission.
+    * @method
+    * @name validatePollSubmission
+    * @param {String} pollId - pollId
+    * @param {String} userId - userId
+    * @returns {String} - message.
+    */
+
+   static validatePollSubmission(pollId= "", userId= "") {
```
Comment (aks30) [inline module/pollSubmissions/helper.js:265]: @deepakr123 - Rename validatePollSubmission to a isPollSubmissionAllowed

File: module/pollSubmissions/helper.js [hunk 30]
```diff
@@ -0,0 +1,315 @@
...
+                throw new Error(messageConstants.apiResponses.USER_ID_REQUIRED_CHECK)
+            }
+
+            let pollDocument = await pollsHelper.validatePoll
+            (
+                pollId
+            )
+            
+            if (!pollDocument.success && !pollDocument.data) {
+               return resolve(pollDocument);
```
Comment (aks30) [inline module/pollSubmissions/helper.js:283]: @deepakr123 - Why a different resolve here ?

From this function - resolve has to be as below 

success: true,
message: messageConstants.apiResponses.POLL_SUBMISSION_VALIDATED,
data: true // True if submission can be made else false

File: module/pollSubmissions/helper.js [hunk 31]
```diff
@@ -0,0 +1,315 @@
...
+                        }
+                    });
+
+                let uniquePolls = [];
+
+                if (result.length > 0) {
+
+                    let mymap = new Map();
+                    uniquePolls = result.filter(el => {
+                        const val = mymap.get(el.name);
```
Comment (aks30) [inline module/pollSubmissions/helper.js:224]: @deepakr123 - We should not use the name as unique, go with id, what I shared was just for reference. Better would be to create gen.utils function which anybody can be used later on. The function takes an array as an argument and the key which should be unique in an array of objects.

File: module/polls/helper.js (hunk 25 above)
Comment (deepakr123) [reply module/polls/helper.js:205]: resolved

File: module/polls/helper.js (hunk 26 above)
Comment (deepakr123) [reply module/polls/helper.js:210]: resolved

File: module/pollSubmissions/helper.js (hunk 28 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:277]: resolved

File: module/pollSubmissions/helper.js (hunk 29 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:265]: resolved

File: module/pollSubmissions/helper.js (hunk 14 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js]: resolved

File: module/polls/helper.js (hunk 24 above)
Comment (deepakr123) [reply module/polls/helper.js:324]: resolved

File: module/polls/helper.js (hunk 26 above)
Comment (deepakr123) [reply module/polls/helper.js:210]: resolved

File: module/pollSubmissions/helper.js (hunk 31 above)
Comment (deepakr123) [reply module/pollSubmissions/helper.js:224]: resolved

File: module/polls/helper.js (hunk 27 above)
Comment (deepakr123) [reply module/polls/helper.js:470]: resolved

**Pull request merged**